src/util.rs

```diff
@@ -39,143 +43,216 @@ pub(crate) fn parse_from_string(input: &str) -> bril_rs::Program {
+/// Write the visualizations to output files in the output directory.
+/// If the directory does not exist, it creates it.
+/// If the directory contains any files whose names conflict with the
+/// output files, it replaces them.
 ///
-/// This struct is intended for use in debugging eggcc. SVG visualizations can
-/// be opened directly in a web browser. Dot visualizations can be converted to
-/// (e.g.) PNGs with utilities like graphviz, e.g.
-/// `dot -Tpng input_cfg.dot -o input_cfg.png`.
-#[allow(unused)]
-pub struct DebugVisualizations {
-    /// The raw CFG before any restructuring, in Dot format,
-    /// for each function in the program.
-    pub(crate) input_cfgs: HashMap<String, String>,
-    /// The restructured CFG ready for RVSDG conversion, in Dot format.
-    pub(crate) restructured_cfgs: HashMap<String, String>,
-    /// The RVSDG, rendered in SVG format.
-    pub(crate) rvsdg: String,
-    /// The PEG, rendered in DOT format.
-    pub(crate) peg: String,
+/// Like other utilities related to `DebugVisualizations`, this method is
+/// only intended for debugging eggcc.
+pub fn visualize(program_path: PathBuf, output_dir: PathBuf) -> io::Result<()> {
+    use std::fs::File;
+    use std::io::Write;
+
+    // make the directory if it doesn't exist
+    if !output_dir.exists() {
+        std::fs::create_dir_all(&output_dir)?;
+    }
+
+    let all_configs = Run::all_configurations_for(program_path);
+
+    let results = all_configs.iter().map(|run| (run, run.run()));
+
+    for (run, result) in results {
+        let mut output_path = output_dir.clone();
+        assert!(output_path.is_dir());
+        output_path.push(format!(
+            "{}{}",
+            run.name(),
+            result.visualization_file_extension
+        ));
+        let mut file = File::create(output_path)?;
+        if let Some(interpreted) = result.result_interpreted {
+            file.write_all(interpreted.as_bytes())?;
+        } else {
+            file.write_all(result.visualization.as_bytes())?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Invokes some program with the given arguments, piping the given input to the program.
+/// Returns an error if the program returns a non-zero exit code.
+/// Code adapted from https://github.com/egraphs-good/egg/blob/e7845c5ae34267256b544c8e6b5bc36d91d096d2/src/dot.rs#L127
+pub fn run_cmd_line<S1, S2, I>(program: S1, args: I, input: &str) -> std::io::Result<String>
+where
+    S1: AsRef<OsStr>,
+    S2: AsRef<OsStr>,
+    I: IntoIterator<Item = S2>,
+{
+    use std::io::Write;
+    use std::process::{Command, Stdio};
+    let mut child = Command::new(program)
+        .args(args)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    let stdin = child.stdin.as_mut().expect("Failed to open stdin");
+    write!(stdin, "{}", input)?;
+
+    let output = child.wait_with_output()?;
+    match output.status.code() {
+        Some(0) => Ok(String::from_utf8(output.stdout).map_err(|e| {
```

Comment:
All these ErrorKind::Others point to it being better for us to have our own error type here, and then return an `anyhow::Result` to capture the io errors as well.

Reply:
I was thinking about that
Would it be better for us to wrap it up in an `EggCCError`? Or we could convert all of our code to use anyhow, which would remove the need for `EggCCError`

Reply:
let's fix this one another time